modules/finality-verifier/src/lib.rs

```diff
@@ -135,17 +137,8 @@ pub mod pallet {
 				<Error<T>>::InvalidAncestryProof
 			);
 
-			// Note that this won't work if we ever change the `ancestry_proof` format to be
-			// sparse since this expects a contiguous set of finalized headers.
-			let _ = T::HeaderChain::append_finalized_chain(ancestry_proof).map_err(|_| {
-				frame_support::debug::error!("Failed to append finalized header chain.");
-				<Error<T>>::FailedToWriteHeader
-			})?;
-
-			frame_support::debug::info!(
-				"Succesfully imported finalized header chain for target header {:?}!",
-				finality_target
-			);
+			T::HeaderChain::append_header(finality_target);
+			frame_support::debug::info!("Succesfully imported finalized header with hash {:?}!", hash);
```

Comment:
It seems to that you're not only "opening the doors to "sparse" proof" here, but also closing the doors to "full" proof here, right? Simply because ancestry proof is always ignored here && you're only "importing" finality target. So unless we'll submit `submit_finality_proof` for every header (which requires GRANDPA justification for every header, which we can't provide), the pallet will always miss some headers. This breaks assumption (at least I was 100% sure in this) that we were going to deploy initial version of bridge with all-headers-in-the-pallet (+historical pruning) && then, eventually. migrate to MMRs (or whatever). Could you confirm that this is an intentional change?

Reply:
Indeed, initially I was more inclined to have full header chain, but it seems it creates more problems than it really solves.

Especially given that:
1. We don't need full header chain for our `MessageLanes` implementations
2. Migrating to MMR would remove full header chain anyway.
3. Pruning is a bit more problematic with full header chain (more data to remove)

Reply:
> It seems to that you're not only "opening the doors to "sparse" proof" here, but also closing the doors to "full" proof here, right?

Yes and no. So _technically_ you can still provide "full" proofs, but now the pallet won't import all the headers from that proof to the base bridge pallet (for the reasons Tomek stated above).